linker/src/lib.rs

```diff
@@ -27,113 +29,156 @@ pub fn link<T: FieldElement>(graph: PILGraph<T>) -> Result<PILFile<T>, Vec<Strin
 
     let mut errors = vec![];
 
-    let pil = graph
-        .objects
+    // Extract the utilities and sort them into namespaces where possible.
+    let mut current_namespace = None;
+    let mut pil = graph
+        .definitions
         .into_iter()
-        .flat_map(|(location, object)| {
-            let mut pil = vec![];
-
-            if let Some(degree) = object.degree {
-                if degree != main_degree {
-                    errors.push(format!(
-                        "Machine {location} should have degree {main_degree}, found {}",
-                        degree
-                    ))
-                }
+        .sorted_by_cached_key(|(namespace, _)| {
+            let mut namespace = namespace.clone();
+            let name = namespace.pop();
+            // This is needed to group into namespaces first.
+            (namespace, name)
+        })
+        .flat_map(|(mut namespace, e)| {
+            let name = namespace.pop().unwrap();
+            let def = PilStatement::LetStatement(0, name.to_string(), Some(e));
+
+            if current_namespace != Some(namespace.clone())
+                && !(current_namespace.is_none() && namespace.is_empty())
+            {
+                current_namespace = Some(namespace.clone());
+                vec![
+                    PilStatement::Namespace(
+                        0,
+                        namespace.relative_to(&AbsoluteSymbolPath::default()),
+                        Expression::Number(T::from(main_degree)),
+                    ),
+                    def,
+                ]
+            } else {
+                vec![def]
+            }
+        })
```

Comment:
This whole block is a bit confusing, can you elaborate more?

Reply:
Actually I was able to simplify the code.